verifier.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratify
+
+import (
+	"context"
+	"fmt"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/registry"
+)
+
+// registeredVerifiers saves the registered verifier factories.
+var registeredVerifiers map[string]func(config CreateVerifierOptions) (Verifier, error)
+
+// Verifier is an interface that defines methods to verify an artifact associated
+// with a subject.
+type Verifier interface {
+	// Name returns the name of the verifier.
+	Name() string
+
+	// Type returns the type name of the verifier.
+	Type() string
+
+	// Verifiable returns if the verifier can verify the given reference.
+	Verifiable(ctx context.Context, referrerDescriptor ocispec.Descriptor) bool
+
+	// Verify verifies the given reference of a subject and returns the verification result
+	Verify(ctx context.Context,
+		subjectReference registry.Reference,
+		referrerDescriptor ocispec.Descriptor,
+		referrerStore Store) (VerifierResult, error)
+}
+
+// VerifierResult defines the verification result that a verifier plugin must return.
+type VerifierResult struct {
+	// Err is the error that occurred when the verification failed.
+	// If the verification is successful, this field should be nil. Optional
+	Err error
+
+	// Description describes the verification result if needed. Optional
+	Description string
+
+	// Verifier refers to the verifier that generated the result. Required.
+	Verifier Verifier
+
+	// Extensions is additional information that can be used to provide more
+	// context about the verification result. Optional.
+	Extensions any
+}
+
+// CreateVerifierOptions represents the options to create a verifier.
+type CreateVerifierOptions struct {
+	// Name is the unique identifier of a verifier instantce. Required.
+	Name string
+	// Type represents a specific implementation of a verifier. Required
+	// Note: there could be multiple verifiers of the same type with different names.
+	Type string
+	// ArtifactTypes is the list of artifact types that the verifier can verify. Required.
+	ArtifactTypes []string
```

Comment:
How do you test if a descriptor is verifiable? purely based on artifact type?

Should `Verifier` expose `ArtifactTypes()`?